# Prediction Of Heart Disease

In [36]:
#importing pandas library and getting the dataset
import pandas as pd
heart_data=pd.read_csv('C:\\Users\\Antra Phukan\\Desktop\\New folder\\heart disease prediction\\heart.csv')

In [8]:
#Information about the data
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


we have 14 columns in our data set

In [11]:
#Total Rows and Columns
heart_data.shape

(1025, 14)

There are 1025 rows and 14 columns

In [12]:
#statiscal description about the data
heart_data.describe()

age          sex           cp     trestbps        chol  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.00000   
mean     54.434146     0.695610     0.942439   131.611707   246.00000   
std       9.072290     0.460373     1.029641    17.516718    51.59251   
min      29.000000     0.000000     0.000000    94.000000   126.00000   
25%      48.000000     0.000000     0.000000   120.000000   211.00000   
50%      56.000000     1.000000     1.000000   130.000000   240.00000   
75%      61.000000     1.000000     2.000000   140.000000   275.00000   
max      77.000000     1.000000     3.000000   200.000000   564.00000   

               fbs      restecg      thalach        exang      oldpeak  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.000000   
mean      0.149268     0.529756   149.114146     0.336585     1.071512   
std       0.356527     0.527878    23.005724     0.472772     1.175053   
min       0.000000     0.000000    71.000000     0.000000     0.000000   
25%       0.000000     0.000000   132.000000     0.000000     0.000000   
50%       0.000000     1.000000   152.000000     0.000000     0.800000   
75%       0.000000     1.000000   166.000000     1.000000     1.800000   
max       1.000000     2.000000   202.000000     1.000000     6.200000   

             slope           ca         thal       target  
count  1025.000000  1025.000000  1025.000000  1025.000000  
mean      1.385366     0.754146     2.323902     0.513171  
std       0.617755     1.030798     0.620660     0.500070  
min       0.000000     0.000000     0.000000     0.000000  
25%       1.000000     0.000000     2.000000     0.000000  
50%       1.000000     0.000000     2.000000     1.000000  
75%       2.000000     1.000000     3.000000     1.000000  
max       2.000000     4.000000     3.000000     1.000000

In [13]:
#we want to know the total nos of people with heart disease and the ones that are not
heart_data['target'].value_counts()

1    526
0    499
Name: target, dtype: int64

In [6]:
#printing ist 5 rows
heart_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

total heart disease patients=526
total patients without heart disease=499

In [15]:
#splitting the input features and target
features=heart_data.drop(columns='target',axis=1)
targets=heart_data['target']

In [16]:
print(features)

      age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0      52    1   0       125   212    0        1      168      0      1.0   
1      53    1   0       140   203    1        0      155      1      3.1   
2      70    1   0       145   174    0        1      125      1      2.6   
3      61    1   0       148   203    0        1      161      0      0.0   
4      62    0   0       138   294    1        1      106      0      1.9   
...   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
1020   59    1   1       140   221    0        1      164      1      0.0   
1021   60    1   0       125   258    0        0      141      1      2.8   
1022   47    1   0       110   275    0        0      118      1      1.0   
1023   50    0   0       110   254    0        0      159      0      0.0   
1024   54    1   0       120   188    0        1      113      0      1.4   

      slope  ca  thal  
0         2   2     3  
1         0   0     3  
2  

In [17]:
print(targets)

0       0
1       0
2       0
3       0
4       0
       ..
1020    1
1021    0
1022    0
1023    1
1024    0
Name: target, Length: 1025, dtype: int64


# Train-Test Split

Splitting the training and testing data into 70:30

In [18]:
#importing required libraries
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(features,targets,test_size=0.3,random_state=2,stratify=targets) 

In [19]:
print(X_train.shape)

(717, 13)


In [23]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 717 entries, 1015 to 244
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       717 non-null    int64  
 1   sex       717 non-null    int64  
 2   cp        717 non-null    int64  
 3   trestbps  717 non-null    int64  
 4   chol      717 non-null    int64  
 5   fbs       717 non-null    int64  
 6   restecg   717 non-null    int64  
 7   thalach   717 non-null    int64  
 8   exang     717 non-null    int64  
 9   oldpeak   717 non-null    float64
 10  slope     717 non-null    int64  
 11  ca        717 non-null    int64  
 12  thal      717 non-null    int64  
dtypes: float64(1), int64(12)
memory usage: 78.4 KB


In [21]:
print(X_test.shape)

(308, 13)


In [22]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 308 entries, 622 to 530
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       308 non-null    int64  
 1   sex       308 non-null    int64  
 2   cp        308 non-null    int64  
 3   trestbps  308 non-null    int64  
 4   chol      308 non-null    int64  
 5   fbs       308 non-null    int64  
 6   restecg   308 non-null    int64  
 7   thalach   308 non-null    int64  
 8   exang     308 non-null    int64  
 9   oldpeak   308 non-null    float64
 10  slope     308 non-null    int64  
 11  ca        308 non-null    int64  
 12  thal      308 non-null    int64  
dtypes: float64(1), int64(12)
memory usage: 33.7 KB


# Training our Model
we train the Machine learning Model using Logistic Regression

In [26]:
from sklearn.linear_model import LogisticRegression
model= LogisticRegression()

# fit the model with data
model.fit(X_train,Y_train)

C:\Users\Antra Phukan\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

# Model Evaluation

In [28]:
#importing necessary function
from sklearn.metrics import accuracy_score
X_train_prediction=model.predict(X_train)
training_data_accuracy=accuracy_score(X_train_prediction,Y_train)
print('Training data accuracy is:',training_data_accuracy)

Training data accuracy is: 0.8647140864714087


Our training data is 86.47% accurate

In [29]:
X_test_prediction=model.predict(X_test)
test_data_accuracy=accuracy_score(X_test_prediction,Y_test)
print('Test data accuracy is:',test_data_accuracy)

Test data accuracy is: 0.7922077922077922


our test data is 79.22% accurate

In [30]:
#building the final predictive system which will give us the target by knowing the input features
input_data=(70, 1, 0, 145, 174, 0, 1, 125, 1, 2.6, 0, 0, 3 )
import numpy as np
input_data_as_numpy_array=np.asarray(input_data)
input_data_reshaped=input_data_as_numpy_array.reshape(1,-1)

prediction=model.predict(input_data_reshaped)
print(prediction)

[0]


In [35]:
if(prediction[0]==0):
  print('The person does not have heart disease')
else:
  print('The person has heart disease')


The person does not have heart disease


# END